contracts/BC_fusion/StakePool.sol

```diff
@@ -0,0 +1,172 @@
+        address sender = _msgSender();
+        require(sender == STAKE_HUB, "NOT_STAKE_HUB");
+        _;
+    }
+
+    modifier onlyValidatorSet() {
+        address sender = _msgSender();
+        require(sender == VALIDATOR_SET, "NOT_VALIDATOR_SET");
+        _;
+    }
+
+    modifier whenNotPaused() {
+        require(!IStakeHub(STAKE_HUB).isPaused(), "CONTRACT_IS_STOPPED");
+        _;
+    }
+
+    /*----------------- external functions -----------------*/
+    function initialize(address _validator, uint256 _selfDelegateAmt) public payable initializer {
+        validator = _validator;
+
+        _bootstrapInitialHolder(_selfDelegateAmt);
+    }
+
+    function delegate(
+        address _delegator,
+        uint256 _bnbAmount
+    ) external payable onlyStakeHub whenNotPaused returns (uint256) {
+        return _stake(_delegator, _bnbAmount);
+    }
+
+    function undelegate(address _delegator, uint256 _sharesAmount) external onlyStakeHub whenNotPaused {
+        require(_sharesAmount != 0, "ZERO_UNDELEGATE");
+        require(_sharesAmount <= _sharesOf(_delegator), "INSUFFICIENT_BALANCE");
+
+        // lock the tokens
+        _transfer(_delegator, address(this), _sharesAmount);
+        _lockedShares[_delegator] += _sharesAmount;
+
+        // add to the queue
+        UnbondRequest memory request =
+            UnbondRequest({amount: _sharesAmount, unlockTime: block.timestamp + IStakeHub(STAKE_HUB).getUnbondTime()});
+        bytes32 hash = keccak256(abi.encodePacked(_delegator, _sharesAmount, block.timestamp));
+        _unbondRequests[hash] = request;
+        _unbondRequestsQueue[_delegator].pushBack(hash);
+
+        emit UnbondRequested(_delegator, _sharesAmount, request.unlockTime);
+    }
+
+    function claim(address _delegator) external onlyStakeHub whenNotPaused returns (uint256) {
+        require(_unbondRequestsQueue[_delegator].length() != 0, "NO_UNBOND_REQUEST");
+
+        bytes32 hash = _unbondRequestsQueue[_delegator].front();
+        UnbondRequest memory request = _unbondRequests[hash];
+        require(block.timestamp >= request.unlockTime, "NOT_UNLOCKED");
+
+        // remove from the queue
+        _unbondRequestsQueue[_delegator].popFront();
+        delete _unbondRequests[hash];
+
+        // unlock and burn the shares
+        _lockedShares[_delegator] -= request.amount;
+        _burn(address(this), request.amount);
+        emit UnbondClaimed(_delegator, request.amount);
+
+        uint256 bnbAmount = getPooledBNBByShares(request.amount);
+        _totalPooledBNB -= bnbAmount;
+        return bnbAmount;
+    }
+
+    function distributeReward(uint256 _amount) external onlyValidatorSet {
+        _totalReceivedReward += _amount;
+        _totalPooledBNB += _amount;
+        emit RewardReceived(_amount);
+    }
+
+    function felony(uint256 _amount) external onlyValidatorSet {
+        uint256 sharesAmount = getSharesByPooledBNB(_amount);
+        _burn(validator, sharesAmount);
```

Comment:
The slashed BNB should to to somewhere else, like system reward contract.

Reply:
Done. As BNB are deposited in `BSCValidatorSet`, we just need to write down here.

Reply:
Done